draft-ietf-jsonpath-base.md

```diff
@@ -413,51 +433,70 @@ the grammar in this document, its UTF-8 form SHOULD first be decoded into
-
-If no node is selected and no error has occurred, an implementation MUST
-return an empty list of nodes.
-
-Syntax errors in the JSONPath SHOULD be detected before selection is attempted
-since these errors do not depend on the JSON value.
-Therefore, an implementation SHOULD take a JSONPath and produce an optional
-syntax error and then,
-if and only if an error was not produced, SHOULD take a JSON value and
-produce a list of nodes or an error (but not both).
-
-Alternatively, an implementation MAY take a JSONPath and a JSON value
-and produce a list of nodes or an optional error (but not both).
-
-For any implementation, if a syntactically invalid JSONPath is provided,
-the implementation MUST return an error.
-
-If a syntactially invalid JSON value is provided, any implementation SHOULD
-return an error.
+A string to be used as a JSONPath query needs to be *well-formed* and
+*valid*.
+A string is a well-formed JSONPath query if it conforms to the syntax
+of JSONPath.
+A well-formed JSONPath query is valid if it also fulfills all semantic
+requirements posed by this document.
+
+Well-formedness and validity of JSONPath queries are independent of
+the JSON value the query is applied to; no further errors can be
+raised during application of the query to a JSON value.
+
+(Obviously, an implementation can still fail when executing a JSONPath
+query, e.g., because of resource depletion, but this is not modeled in
+the present specification.)
+
+> Discussion: are we ready for the error model defined here, i.e. one
+> that does not have runtime errors, but only compile time errors?
+
+## Processing Model
+
+A valid JSONPath query is executed against a JSON value, the *Argument*, and
+produces a list of zero or more nodes of the JSON value which are
+selected by the JSONPath.
+(Note that the term JSON value also implies that this value complies
+to the definitions in {{-json}}, i.e., is indeed a JSON value.)
+
+The JSONPath query is a sequence of zero or more *selectors*, each of
+which is applied to the result of the previous selector and leaves
+input to the next selector.
+These results and inputs take the form of a *nodelist*, i.e., a
+sequence of zero or more nodes.
+
+The nodelist going into the first selector contains a single node,
+holding the Argument.
+The nodelist resulting from the last selector is presented as the
+result of the query; depending on the specific API, it might be
+presented as an array of the JSON values at the nodes, an array of
+Output Paths referencing the nodes, or both — or some other
+representation as desired by the implementation.
+Note that the API must be capable of presenting an empty nodelist as
+the result of the query.
+
+A selector performs its function on each of the nodes in its input
+nodelist, during such a function execution, such a node is referred to
+as "current node".  Each of these function executions produces a
+result nodelist, which are then combined by algorithm *combine1* into
```

Comment:
combine1 appears to be undefined.

Reply:
Sure -- it has the strange name because it is a placeholder that needs to be filled in based on the discussion points.  But we can live with this placeholder while we resolve the discussions, hence giving it a name is a good idea for these discussions.